### PR TITLE
feat(payments): INT-3610 added Mollie Checkout

### PR DESCRIPTION
**Pull request by JoseAlvarezHdz**
## What? [INT-3610](https://jira.bigcommerce.com/browse/INT-3610)
Create strategy for Mollie payment method on checkout

## Why?
In order to be able to purchase through Mollie

## Testing / Proof

[Video](https://drive.google.com/file/d/1ZYGejKPOXk_jhI99w0pdcIA_vIFd7FwN/view?usp=sharing)

screen-recorder-wed-dec-30-2020-11-32-54 (1)

## Siblings

[JS](https://github.com/bigcommerce/checkout-js/pull/490)

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/payment/payment-request-options.ts [hunk 1]
```diff
@@ -168,4 +169,10 @@ export interface PaymentInitializeOptions extends PaymentRequestOptions {
      * They can be omitted unless you need to support StripeV3.
      */
     stripev3?: StripeV3PaymentInitializeOptions;
+
```
Comment (vmparra) [inline src/payment/payment-request-options.ts:172]: StripeV3?

File: src/payment/payment-strategy-type.ts [hunk 2]
```diff
@@ -42,6 +42,7 @@ enum PaymentStrategyType {
     SEZZLE = 'sezzle',
     ZIP = 'zip',
     CONVERGE = 'converge',
+    MOLLIE = 'mollie',
```
Comment (vmparra) [inline src/payment/payment-strategy-type.ts:45]: Can you please sort this options alphabetically 💅🏻

File: src/payment/strategies/mollie/mollie-payment-strategy.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        }
+
+        return this._mollieClient;
+    }
+
+    private _mountElements() {
+        const mollieClient = this._getMollieClient();
+        const { cardNumberId, cardCvcId, cardExpiryId, cardHolderId} = this._getInitializeOptions();
+
+        this._cardHolderElement = mollieClient.createComponent('cardHolder', {
```
Comment (vmparra) [inline src/payment/strategies/mollie/mollie-payment-strategy.ts:141]: Can we send this options from the frontend?

**Review by vmparra — CHANGES_REQUESTED** [review]
Thanks @JoseAlvarezHdz for doing this, so far it looks real good. I just left a few comments for you. Please let me know if you have questions.

File: src/payment/payment-strategy-type.ts (hunk 2 above)
Comment (JoseAlvarezHdz) [reply src/payment/payment-strategy-type.ts:45]: You mean like all the gateway names?

File: src/payment/payment-request-options.ts (hunk 1 above)
Comment (JoseAlvarezHdz) [reply src/payment/payment-request-options.ts:172]: oh no, i added the mollie option down here

File: src/payment/strategies/mollie/mollie-payment-strategy.ts (hunk 3 above)
Comment (JoseAlvarezHdz) [reply src/payment/strategies/mollie/mollie-payment-strategy.ts:141]: what do you mean from the frontend?

Comment (JoseAlvarezHdz) [reply src/payment/strategies/mollie/mollie-payment-strategy.ts:141]: you mean the component type?

**Comment by JoseAlvarezHdz** [comment]
@vmparra i did the changes you mentions, can you validate them please?

File: src/payment/strategies/mollie/mollie-payment-strategy.ts (hunk 3 above)
Comment (vmparra) [reply src/payment/strategies/mollie/mollie-payment-strategy.ts:141]: No, I meant the css classes.

File: src/payment/payment-request-options.ts (hunk 1 above)
Comment (vmparra) [reply src/payment/payment-request-options.ts:172]: Look at the documentation.

File: src/payment/payment-strategy-type.ts (hunk 2 above)
Comment (vmparra) [reply src/payment/payment-strategy-type.ts]: Correct

File: src/payment/payment-request-options.ts (hunk 1 above)
Comment (JoseAlvarezHdz) [reply src/payment/payment-request-options.ts:172]: my bad, now is fix

File: src/payment/strategies/mollie/mollie-payment-strategy.ts (hunk 3 above)
Comment (JoseAlvarezHdz) [reply src/payment/strategies/mollie/mollie-payment-strategy.ts:141]: they are gone now

**Comment by JoseAlvarezHdz** [comment]
@vmparra i did the changes you mention, can you validate them please?

**Review by vmparra — APPROVED** [review]
LGTM 👍🏻

File: src/payment/payment-methods.mock.ts [hunk 4]
```diff
@@ -606,6 +606,36 @@ export function getCheckoutcom(): PaymentMethod {
...
+            isVaultingCvvEnabled: false,
+            isVaultingEnabled: false,
+            isVisaCheckoutEnabled: false,
+            merchantId: 'test_T0k3n',
+            requireCustomerCode: false,
+            testMode: true,
+        },
+        initializationData: null,
+        type: 'PAYMENT_TYPE_API',
+        };
```
Comment (capsula4) [inline src/payment/payment-methods.mock.ts:636]: minor, indentation off

File: src/payment/strategies/mollie/mollie-payment-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        const { token, error } = await this._getMollieClient().createToken();
+
+        if (error) {
+            return Promise.reject(error);
+        }
+
+        return this._store
+                .dispatch(this._orderActionCreator.submitOrder(order, options))
+                .then(() => {
```
Comment (capsula4) [inline src/payment/strategies/mollie/mollie-payment-strategy.ts:75]: prefer async/await over then/catch

File: src/payment/strategies/mollie/mollie-script-loader.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export default class MollieScriptLoader {
+    constructor(
+        private _scriptLoader: ScriptLoader,
+        private _window: MollieHostWindow = window
+    ) {}
+
+    load(merchantId: string, locale: string, testmode: boolean): Promise<MollieClient> {
+        return this._scriptLoader
+            .loadScript('https://js.mollie.com/v1/mollie.js')
+            .then(() => {
```
Comment (capsula4) [inline src/payment/strategies/mollie/mollie-script-loader.ts:16]: prefer async/await over then/catch

File: src/payment/strategies/mollie/mollie.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    error: object;
+}
+
+export interface MollieConfigurationOptions {
+    locale: string;
+    testmode: boolean;
+}
+
+export interface MollieHostWindow extends Window {
+    Mollie?(
```
Comment (capsula4) [inline src/payment/strategies/mollie/mollie.ts:17]: `Mollie?` should this not be optional?

File: src/payment/strategies/mollie/mollie-script-loader.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+                if (!this._window.Mollie) {
```
Comment (capsula4) [inline src/payment/strategies/mollie/mollie-script-loader.ts:17]: probably you can use a custom typeguard here instead
```
if (!isMollieWindow(this._window)) {
  // throw
}

```

Comment (JoseAlvarezHdz) [reply src/payment/strategies/mollie/mollie-script-loader.ts:17]: if i add this typeguard for some reason i get this warning

_**Cannot invoke an object which is possibly 'undefined'.**_

File: src/payment/strategies/mollie/mollie.ts (hunk 7 above)
Comment (JoseAlvarezHdz) [reply src/payment/strategies/mollie/mollie.ts:17]: if i remove the optional sign i get a warning 

_**Property 'Mollie' is missing in type 'Window & typeof globalThis' but required in type 'MollieHostWindow'.ts(2741)**_

**Comment by JoseAlvarezHdz** [comment]
@capsula4  i did some of the changes you mention, can you validate them please?

File: src/payment/strategies/mollie/mollie-script-loader.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+export default class MollieScriptLoader {
+    constructor(
+        private _scriptLoader: ScriptLoader,
+        private _window: MollieHostWindow = window
+    ) {}
+
+    async load(merchantId: string, locale: string, testmode: boolean): Promise<MollieClient> {
+        await this._scriptLoader.loadScript('https://js.mollie.com/v1/mollie.js');
+
+        if (!this._window.Mollie) {
```
Comment (capsula4) [inline src/payment/strategies/mollie/mollie-script-loader.ts:16]: ideally the best way to check for this is using a typeguard as suggested before.

This is an example

https://github.com/bigcommerce/checkout-sdk-js/blob/master/src/analytics/is-analytics-step-tracker-window.ts
https://github.com/bigcommerce/checkout-sdk-js/blob/master/src/analytics/analytics-tracker-window.ts

Notice the AnalyticsTracker is not an optional attribute of the `AnalyticsTrackerWindow` interface, which is what I suggested for the `MolllieWindow` interface

**Review by capsula4 — APPROVED** [review]

**Pull request merged**